### ISSUE TITLE
UI: delete_entity - remove assertion search after entity deletion

### ISSUE THREAD

**Issue by lpramuk** [SatelliteQE/robottelo#3104]
I'm not quite fine with following lines of our code  https://github.com/SatelliteQE/robottelo/blob/master/robottelo/ui/base.py#L209-L226

Such assert should be in tests side.
Simply this goes against object model if we assert on [tested object] side within its methods.

- Either we remove it here and implement within tests, but then we would lose single point of verification logic
- Or we can have common helpers on [test] side for that.
```
[tested object] <--- [test]
.delete()                 .assert_deleted()
```
At least it is not consistent with create() where we don't search for entity after its creation
or with update() where we don't check if property gets properly updated... it is left for test logic.

I though of this during investigation of failed ```test_positive_delete_bz1225588```, where _non-empty_ organizations don't get deleted immediatelly...we are "asserting" right away.

Ideas/comments :-)

**Comment by rplevka** [comment]
+1
what's more, I believe that we should handle the situations, as operations (delete, create, or whatever) aren't completed immediately as we are being notified (by ui notification) as a bug.

What we do instead is that we implement a workaround directly to the core of our 'object representation', masking the real problem.
I would strictly proceed with filing a bug, putting a `if_bz_open` decorator to the 'workaround loop' __inside tests__

**Comment by oshtaier** [comment]
-1
We have very specific reasons to have current implementation in a way that it is. And, of course them applied only for delete() operation and should not be applied on create() or update() operation due obvious logic hidden in a way of non-existent element verification. Basically, it is not so easy to check that element is actually deleted without having unnecessary waiters/delays. Of course, other options were also covered and decided to go with that approach as it seems most optimal which of course cover most (not all) cases

>Either we remove it here and implement within tests, but then we would lose single point of verification logic

Yeah and add hundreds lines of code, change architecture and spend tons of hours to re-test all functionality. I think it is very-very bad idea. Actually, we used completely opposite approach and removed all logic from tests for a reason.

>Or we can have common helpers on [test] side for that.

Also seems as waste of time and efforts. And actually, I don't see how that can be easily implemented or what will be the difference with current approach.

`test_positive_delete_bz1225588` is a corner case, so please use a workaround for it. +You can fill a defect as suggested by @rplevka and add corresponding decorator as organization should be deleted in the 10-15 seconds that we are using in our verification. Otherwise it is a defect as user will never know that he should for such period of time

And now, let's get to the core of your question. What is specific problem with current approach?
That one?
>Such assert should be in tests side.
Simply this goes against object model if we assert on [tested object] side within its methods.

Yes, it is well-known cons of that approach, that we accepted to have better logic/performance/code size.

But anyway, if you like to have better one, please proceed with your own PR.

**Comment by elyezer** [comment]
I searched ui/base.py for assert and I found nothing. @lpramuk is this issue still valid?

**Comment by rplevka** [comment]
@elyezer, it's not an assert command, but the logic contained in a block of code @lpramuk highlighted (https://github.com/SatelliteQE/robottelo/blob/master/robottelo/ui/base.py#L209-L226).
We are applying a workaround directly in the code which might go against the design, introducing a non-consistency into the code logic.

**Comment by oshtaier** [comment]
@elyezer @rplevka 
Right, we performed that implementation (which, of course, was agreed with a team) to have huge performance boost and remove a lot of redundant code.
It is not workaround, but a proper solution due selenium logic and other reasons which were discussed a lot.
>might go against the design, introducing a non-consistency into the code logic.
Yes, it breaks logic a little, but that was bearable risk.

Anyway, I said that above, so please read that information once more. If anyone has better idea or time, please submit PR or ask for help and I will be more than happy to join any efforts in that direction as I know all UI code on a really nice level.

**Comment by elyezer** [comment]
That retries is needed because some async tasks that Satellite creates, I don't see it as a problem, but as a way to deal with the asynchronous behavior.

**Comment by abalakh** [comment]
+1 to @oshtaier & @elyezer here. `delete()` may appear to have a bit inconsistent design with `create()` or `update()`, but we had to do that because `delete()` is not atomic operation and often takes some time. Otherwise we'd had to add a workaround with loops and waiters inside each delete test or use some helper each time... and hope that we won't forget to add it in some future to new tests otherwise we'd receive some random failures.
From the robottelo user perspective I don't wanna know all the technical details on how sattelite handles such operations on low-level, I just want to use `delete()` and expect that entity was deleted and I can proceed with any next operation/assertion i'd like to. Thoughts like 'oh, I used `delete()`, maybe i should wait a bit just in case' shouldn't bother me. And, AFAIK, tests usually don't use waiters, most async operations are handled by framework to make things more user-friendly.
Long story short, I completely agree with the point:
> I don't see it as a problem, but as a way to deal with the asynchronous behavior.

:)

**Comment by elyezer** [comment]
With all that said, I think we can close this issue.

**Comment by oshtaier** [comment]
Right, we can close that issue as initial thoughts described here are opposite to chosen approach.

Anyway, we still have https://github.com/SatelliteQE/robottelo/issues/3101 issue where we can continue tracking for any improvements needed for delete functionality
